### PR TITLE
new: add support for og:image/twitter:image meta tags and img tags for image preview

### PR DESCRIPTION
**Pull request by GamingGeek**
Resolves Essential's image links not previewing, allows many other sources to be previewed

### CODE REVIEW

File: src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java [hunk 1]
```diff
@@ -36,6 +40,8 @@ public class ImagePreview {
     private int tex = -1;
     private int width = 100;
     private int height = 100;
+    private static final Pattern ogpImageRegex = Pattern.compile("<meta property=\"(?:og:image|twitter:image)\" content=\"(?<url>.+?)\".*?/?>");
```
Comment (KodingDev) [inline src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java:43]: Top of the file, all caps for both.

File: src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java [hunk 2]
```diff
@@ -148,11 +154,38 @@ private void loadUrl(String url) {
             connection = (HttpURLConnection) u.openConnection();
             connection.setRequestMethod("GET");
             connection.setUseCaches(true);
+            connection.setInstanceFollowRedirects(true);
             connection.addRequestProperty("User-Agent", "Patcher Image Previewer");
             connection.setReadTimeout(15000);
             connection.setConnectTimeout(15000);
             connection.setDoOutput(true);
 
+            if (connection.getHeaderField("Content-Type").contains("text/html")) {
```
Comment (KodingDev) [inline src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java:163]: Should be equals.

File: src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java [hunk 3]
```diff
@@ -148,11 +154,38 @@ private void loadUrl(String url) {
...
+                BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
+                String currentLine, imageURL = "";
+                Matcher matcher = null;
+                while ((currentLine = reader.readLine()) != null) {
+                    if (currentLine.contains("og:image") || currentLine.contains("twitter:image"))
```
Comment (KodingDev) [inline src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java:168]: Use brackets.

File: src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java [hunk 4]
```diff
@@ -148,11 +154,38 @@ private void loadUrl(String url) {
...
+                        matcher = ogpImageRegex.matcher(currentLine);
+                    else if (currentLine.contains("<img ")) matcher = imgTagRegex.matcher(currentLine);
```
Comment (KodingDev) [inline src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java:170]: Should match for both, contains statements are unneeded.

File: src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java [hunk 5]
```diff
@@ -148,11 +154,38 @@ private void loadUrl(String url) {
...
+                        try {
+                            imageURL = matcher.group("url");
+                            if (imageURL.startsWith("/")) {
+                                URL urlObj = new URL(url);
+                                imageURL = urlObj.getProtocol() + "://" + urlObj.getHost() + imageURL;
+                            }
+                            break;
+                        } catch (Exception ignored) {}
+                    }
+                    matcher = null;
```
Comment (KodingDev) [inline src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java:181]: Unnecessary.

**Review by KodingDev — CHANGES_REQUESTED** [review]

File: src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java (hunk 2 above)
Comment (GamingGeek) [reply src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java:163]: Equals won't work if the header contains encoding

Comment (KodingDev) [reply src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java:163]: Possibly replace with startsWith?

**Review by KodingDev — CHANGES_REQUESTED** [review]
ayo what the formatting do brah

**Review by KodingDev — APPROVED** [review]

File: src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java [hunk 6]
```diff
@@ -148,11 +154,33 @@ private void loadUrl(String url) {
...
             connection.setRequestMethod("GET");
             connection.setUseCaches(true);
+            connection.setInstanceFollowRedirects(true);
             connection.addRequestProperty("User-Agent", "Patcher Image Previewer");
             connection.setReadTimeout(15000);
             connection.setConnectTimeout(15000);
             connection.setDoOutput(true);
 
+            if (connection.getHeaderField("Content-Type").contains("text/html")) {
+                String body = IOUtils.toString(connection.getInputStream());
```
Comment (Sychic) [inline src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java:164]: Shouldn't this be in the try block? I would assume the other `connection.getInputStream()` is in one for a reason.

Comment (GamingGeek) [reply src/main/java/club/sk1er/patcher/screen/render/overlay/ImagePreview.java]: I moved the try block to contain it all and remove the duplicate getInputStream calls

**Review by Sychic — APPROVED** [review]

**Pull request merged**